### ISSUE TITLE
Export the "mixins". Show how to export them in the proper order and document it.

### ISSUE THREAD

**Issue by mwilliams2014** [liferay/alloy-editor#1314]

**Comment by wincent** [comment]
Labeling this as a "feature" although we could just as easily classify it as a bug because it is something you used to be able to do straightforwardly in 1.x and it became difficult or impossible in 2.x.

**Comment by wincent** [comment]
For the purposes of documentation, these are the existing wrapping patterns that you can see in 2.x:

```
ButtonActionStyle
  ButtonStateClasses
    ButtonStyle
      Component (x3)

ButtonActionStyle
  ButtonStyle
    Component

ButtonCfgProps
  ButtonKeystroke
    ButtonStateClasses
      Component

ButtonCfgProps
  WidgetDropdown
    WidgetFocusManager
      Component

ButtonCommand
  Component (x2)

ButtonCommand
  ButtonCommandActive
    ButtonStateClasses
      Component (x9)

ButtonCommand
  ButtonKeystroke
    ButtonStateClasses
      ButtonStyle
        Component (x3)

ButtonCommand
  ButtonStateClasses
    ButtonStyle
      Component (x5)

ButtonCommand
   ButtonStyle
     Component

ButtonKeystroke
  Component

ButtonProps
  ButtonStateClasses
    Component (x4)

ButtonProps
  ButtonStateClasses
    ButtonKeystroke
      Component

ButtonProps
  ButtonStateClasses
    ButtonStyle
      Component

ButtonStateClasses
  Component

WidgetArrowBox
  WidgetDropdown
    WidgetExclusive
      WidgetFocusManager
        WidgetPosition
          ToolbarButtons
            Component (x2)

WidgetExclusive
  WidgetFocusManager
    Component

WidgetFocusManager
  Component (x5)
```

Some but not all of them are order-sensitive (a terrible property for them to have, but fixing it basically requires a total rewrite). So looking at that list, we can simplify it to something like the following, but will need to dig though the code to determine what the real dependencies are (which is definitely a black mark against the architectural design):

```
ButtonActionStyle
ButtonStateClasses
ButtonStyle

ButtonCfgProps
ButtonKeystroke
ButtonStateClasses

ButtonCfgProps
WidgetDropdown
WidgetFocusManager

ButtonCommand
ButtonCommandActive
ButtonStateClasses
ButtonStyle

ButtonCommand
ButtonKeystroke (or ButtonStateClasses)
ButtonStateClasses (or ButtonKeystroke)

ButtonProps

WidgetArrowBox
WidgetDropdown
WidgetExclusive
WidgetFocusManager
WidgetPosition
ToolbarButtons
```

**Comment by wincent** [comment]
@mwilliams2014: hoping this is enough information to be able to prepare documentation for creating custom buttons with AlloyEditor 2.x

My understanding is that we don't want to document the `AlloyEditor.Compat` stuff yet, because we don't know if anybody will ever need it, but we could document how to make a custom button in 2.x using `AlloyEditor.Base`. I also suspect that we may not need to document the `Widget*` stuff (or `ToolbarButtons`), as I suspect that most use cases are covered by the `Button*` HOCs.

The information here may not exactly coincide with what you need (ie. it may be too much in some places, and not enough in others), but hopefully it is not too far off the mark.

---

# Writing custom buttons with AlloyEditor 1.x

AlloyEditor 1.x bundled and used an older version of React (v0.14.2). Buttons and toolbars were created using `React.createClass` plus [a series of "mixins"](https://github.com/liferay/alloy-editor/tree/1.x/src/ui/react/src/components/base) that were available on the global `AlloyEditor` object:

-   `AlloyEditor.ButtonActionStyle`
-   `AlloyEditor.ButtonCommandActive`
-   `AlloyEditor.ButtonCommand`
-   `AlloyEditor.ButtonKeystroke`
-   `AlloyEditor.ButtonCfgProps`
-   `AlloyEditor.ButtonStateClasses`
-   `AlloyEditor.ButtonStyle`
-   `AlloyEditor.ToolbarButtons`
-   `AlloyEditor.WidgetArrowBox`
-   `AlloyEditor.WidgetDropdown`
-   `AlloyEditor.WidgetExclusive`
-   `AlloyEditor.WidgetFocusManager`
-   `AlloyEditor.WidgetInteractionPoint`
-   `AlloyEditor.WidgetPosition`

The mixins are a vehicle for sharing common behavior across components by adding properties and methods to their targets. You can find examples of the mixins being used in [the buttons](https://github.com/liferay/alloy-editor/tree/1.x/src/ui/react/src/components/buttons) and [toolbars](https://github.com/liferay/alloy-editor/tree/1.x/src/ui/react/src/components/toolbars) that come with AlloyEditor, and presumably, people who made custom buttons were probably using them as well.

I won't document them here because I don't have any direct experience with 1.x and I presume that we already have documentation for that somewhere, but the basic usage pattern looks like this:

```javascript
const SomeButton = React.createReactClass({
	mixins: [
		AlloyEditor.ButtonStyle,
		AlloyEditor.ButtonStateClasses,
		AlloyEditor.ButtonActionStyle,
	],

	// Rest of button implementation follows here...
});
```

# Writing custom buttons with Alloy Editor 2.x

AlloyEditor 2.x uses a much newer version of React (v16.8.5) and all the buttons have been rewritten without `React.createClass`; that means they are all now ES6 classes or functional components.

There is no "mixin" equivalent for ES6 classes, and the main abstraction for sharing behaviors between functional components ([hooks](https://reactjs.org/docs/hooks-intro.html)) didn't exist when AlloyEditor 2.x came out, so we replaced the 1.x mixins with "Higher-Order Components" (hereafter, "HOCs"). These are functions which take a component as a parameter, and return a modified component. (Internally, as an implementation detail, these HOCs are actually creating subclasses which overwrite methods and properties on the parent components.)

For 2.x, we consolidated some of the mixins which always ended up being used together into single HOCs, so the list of HOCs is somewhat shorter than the list of mixins that we had in 1.x:

-   `AlloyEditor.Base.ButtonActionStyle`
-   `AlloyEditor.Base.ButtonCommandActive`
-   `AlloyEditor.Base.ButtonCommand`
-   `AlloyEditor.Base.ButtonKeystroke`
-   `AlloyEditor.Base.ButtonProps`
-   `AlloyEditor.Base.ButtonStateClasses`
-   `AlloyEditor.Base.ButtonStyle`
-   `AlloyEditor.Base.ToolbarButtons`
-   `AlloyEditor.Base.WidgetDropdown`
-   `AlloyEditor.Base.WidgetExclusive`
-   `AlloyEditor.Base.WidgetFocusManager`

ie. `WidgetArrowBox`, `WidgetInteractionPoint`, and `WidgetPosition` were all folded into `ToolbarButtons`. Also note that `ButtonCfgProps` was renamed to `ButtonProps`.

The basic usage pattern looks as follows now; instead of listing `mixins`, we pass the component (`SomeButton` in this example) into a HOC (eg. `ButtonStyle`), and then pass _that_ into the next HOC until we've got the desired result:

```javascript
export default ButtonActionStyle(ButtonStateClasses(ButtonStyle(SomeButton)));
```

See [the buttons](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons) and [toolbars](https://github.com/liferay/alloy-editor/tree/master/src/components/toolbars) directories for many usage examples.

## HOC details

HOCs may require other HOCs, or at least the presence of specific methods defined on components, to function properly. Here are some caveats:

### `ButtonActionStyle`

-   **Provides:** `applyStyle()` method.
-   **Requires:** `getStyle()` and `isActive()` methods, which may be obtained from the `ButtonStyle` HOC.
-   **Examples:**
    -   Buttons which all use both `ButtonActionStyle` and `ButtonStyle`:
        -   [`ButtonCode`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-code.jsx)
        -   [`ButtonH1`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-h1.jsx)
        -   [`ButtonH2`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-h2.jsx)
        -   [`ButtonStylesListItem`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-styles-list-item.jsx)
    -   Buttons which implement their own `isActive()` methods, but don't use `ButtonActionStyle` or `ButtonStyle`:
        -   [`ButtonLink`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link.jsx)
        -   [`ButtonTwitter`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-twitter.jsx)

### `ButtonCommand`

-   **Provides:** `execCommand()` method.
-   **Requires:** `command` prop (either explicitly passed or via default props).
-   **Examples:**
    -   See buttons which call `execCommand()` from inside click handlers, or via the `ButtonKeystroke` HOC in response to a key press.

### `ButtonCommandActive`

-   **Provides:** `isActive()` method.
-   **Requires:** `command` prop.
-   **Examples:**
    -   [`ButtonImageAlignCenter`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-image-align-center.jsx)
    -   [`ButtonImageAlignLeft`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-image-align-left.jsx)
    -   [`ButtonImageAlignRight`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-image-align-right.jsx)
    -   [`ButtonIndentBlock`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-indent-block.jsx)()
    -   [`ButtonOutdentBlock`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-outdent-block.jsx)
    -   [`ButtonParagraphAlignLeft`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-paragraph-align-left.jsx)
    -   [`ButtonParagraphAlignRight`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-paragraph-align-right.jsx)
    -   [`ButtonParagraphCenter`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-paragraph-center.jsx)
    -   [`ButtonParagraphJustify`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-paragraph-justify.jsx)
    -   There are no buttons which use `ButtonCommandActive` _and_ `ButtonStyle` (which also provides `isActive()`).

### `ButtonKeystroke`

-   **Provides:** registration/deregistration of key events with CKEditor during `componentDidMount()`/`componentsWillUnmount()`.
-   **Requires:** `keystroke` prop with `fn` (function name) and `keys` properties.
-   **Examples:**
    -   Buttons which rely on `execCommand()`, via the `ButtonCommand` HOC:
        -   [`ButtonBold`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-bold.jsx)
        -   [`ButtonItalic`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-italic.jsx)
        -   [`ButtonUnderline`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-underline.jsx)
    -   Buttons which rely on `_requestExclusive()`, defined locally:
        -   [`ButtonEmbed`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-embed.jsx)
        -   [`ButtonLink`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link.jsx)
        -   [`ButtonLinkBrowse`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link-browse.jsx)

### `ButtonProps`

-   **Provides:** `mergeButtonCfgProps()` method.
-   **Requires:** nothing.
-   **Examples:**
    -   [`ButtonLink`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link.jsx)
    -   [`ButtonLinkBrowse`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link-browse.jsx)

### `ButtonStateClasses`

-   **Requires:** `isActive()` implementation
-   **Examples:**
    -   Buttons with get `isActive()` via the `ButtonComandActive` HOC:
        -   [`ButtonImageAlignCenter`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-image-align-center.jsx)
        -   [`ButtonImageAlignLeft`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-image-align-left.jsx)
        -   [`ButtonImageAlignRight`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-image-align-right.jsx)
        -   [`ButtonIndentBlock`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-indent-block.jsx)()
        -   [`ButtonOutdentBlock`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-outdent-block.jsx)
        -   [`ButtonParagraphAlignLeft`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-paragraph-align-left.jsx)
        -   [`ButtonParagraphAlignRight`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-paragraph-align-right.jsx)
        -   [`ButtonParagraphCenter`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-paragraph-center.jsx)
        -   [`ButtonParagraphJustify`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-paragraph-justify.jsx)
    -   Buttons with get `isActive()` via the `ButtonStyle` HOC:
        -   [`ButtonBold`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-bold.jsx)
        -   [`ButtonCode`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-code.jsx)
        -   [`ButtonH1`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-h1.jsx)
        -   [`ButtonH2`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-h2.jsx)
        -   [`ButtonItalic`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-italic.jsx)
        -   [`ButtonOrderedList`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-ol.jsx)
        -   [`ButtonQuote`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-quote.jsx)
        -   [`ButtonStrike`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-strike.jsx)
        -   [`ButtonStylesListItem`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-styles-list-item.jsx)
        -   [`ButtonSubscript`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-subscript.jsx)
        -   [`ButtonSuperscript`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-superscript.jsx)
        -   [`ButtonUnderline`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-underline.jsx)
        -   [`ButtonUnorderedList`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-ul.jsx)
    -   Buttons that have their own implementation of `isActive()`:
        -   [`ButtonLink`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link.jsx)
        -   [`ButtonTwitter`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-twitter.jsx)

### `ButtonStyle`

-   **Provides:** `getStyle()` and `isActive()` methods.
-   **Requires:** `style` prop.
-   **Examples:** see buttons which use the `ButtonStateClasses` HOC, and/or use the `ButtonActionStyle` HOC's `applyStyle()` method.

### `WidgetDropDown`

-   **Provides:**
    -   `mergeDropdownProps()` and `toggletDropdown()` methods
    -   `expanded` and `tabIndex` props.
-   **Examples:**
    -   `mergeDropdownProps()` is used by the `ToolbarButtons` HOC (ie. implicit dependency) and the [`ButtonLinkEdit`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link-edit.jsx) component.
    -   `toggleDropdown()` is used by many components, via props:
        -   [`ButtonBackgroundColor`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-background-color.jsx)
        -   [`ButtonColor`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-color.jsx)
        -   [`ButtonLinkAutocompleteList`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link-autocomplete-list.jsx)
        -   [`ButtonLinkEdit`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link-edit.jsx)
        -   [`ButtonLinkTargetEdit`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link-target-edit.jsx)
        -   [`ButtonParagraphAlign`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-paragraph-align.jsx)
        -   [`ButtonSpacing`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-spacing.jsx)
        -   [`ButtonStyles`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-styles.jsx)
        -   [`ButtonStylesList`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-styles-list.jsx)
        -   [`ButtonTableCell`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-table-cell.jsx)
        -   [`ButtonTableColumn`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-table-column.jsx)
        -   [`ButtonTableHeading`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-table-heading.jsx)
        -   [`ButtonTableRow`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-table-row.jsx)
    -   It's also used by the bridge modules:
        -   [`MenuButton`](https://github.com/liferay/alloy-editor/blob/master/src/components/uibridge/menu-button.jsx)
        -   [`PanelMenuButton`](https://github.com/liferay/alloy-editor/blob/master/src/components/uibridge/panel-menu-button.jsx)
        -   [`Richcombo`](https://github.com/liferay/alloy-editor/blob/master/src/components/uibridge/richcombo.jsx)

### `WidgetExclusive`

-   **Provides:**
    -   `cancelExclusive()`, `filterExclusive()`, `mergeExclusiveProps()`, `requestExclusive()` methods.
    -   `renderExlusive` prop.
-   **Examples:**
    -   `mergeExclusiveProps()` is used in:
        -   [main.jsx](https://github.com/liferay/alloy-editor/blob/master/src/components/main.jsx)
        -   the `ToolbarButtons` HOC.
    -   `cancelExclusive()` is used in:
        -   [ButtonAccessiblityImageAlt`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-accessibility-image-alt.jsx)
        -   [ButtonCameraImage`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-camera-image.jsx)
        -   [ButtonEmbedEdit`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-embed-edit.jsx)
        -   [ButtonEmbedVideoEdit`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-embed-video-edit.jsx)
        -   [ButtonLinkEdit`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link-edit.jsx)
        -   [ButtonTableEdit`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-table-edit.jsx)
    -   `renderExclusive` is used in:
        -   [ButtonAccessibilityImageAlt`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-accessibility-image-alt.jsx)
        -   [ButtonCamera`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-camera.jsx)
        -   [ButtonEmbed`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-embed.jsx)
        -   [ButtonEmbedEdit`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-embed-edit.jsx)
        -   [ButtonEmbedVideo`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-embed-video.jsx)
        -   [ButtonEmbedVideoEdit`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-embed-video-edit.jsx)
        -   [ButtonLink`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link.jsx)
        -   [ButtonLinkBrowse`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link-browse.jsx)
        -   [ButtonLinkEdit`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link-edit.jsx)
        -   [ButtonTable`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-table.jsx)
        -   [ToolbarAdd`](https://github.com/liferay/alloy-editor/blob/master/src/components/toolbars/toolbar-add.jsx)
    -   `requestExclusive()` is used in:
        -   [ButtonAccessibilityImageAlt`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-accessibility-image-alt.jsx)
        -   [ButtonCamera`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-camera.jsx)
        -   [ButtonEmbed`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-embed.jsxhttps://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-embed.jsx)
        -   [ButtonEmbedVideo`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-embed-video.jsx)
        -   [ButtonLink`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link.jsx)
        -   [ButtonLinkBrowse`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link-browse.jsx)
        -   [ButtonTable`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-table.jsx)
        -   [`ToolbarAdd`](https://github.com/liferay/alloy-editor/blob/master/src/components/toolbars/toolbar-add.jsx)

### `WidgetFocusManager`

-   **Provides:** `focus()`, `handleKey()`, `moveFocus()` methods.
-   **Requires:** `circular`, `focusFirstChild`, `keys`, `onDismiss` props.
-   **Examples:**
    -   [main.jsx](https://github.com/liferay/alloy-editor/blob/master/src/components/main.jsx) (UI) uses `focus`, `handleKey` (on keydown on toolbars), and also uses the `WidgetExclusive` HOC.
    -   These uses `focus`, `handleKey`
        -   [`ToolbarStyles`](https://github.com/liferay/alloy-editor/blob/master/src/components/toolbars/toolbar-styles.jsx)
        -   [`ToolbarAdd`](https://github.com/liferay/alloy-editor/blob/master/src/components/toolbars/toolbar-add.jsx)
        -   [`ButtonDropdown`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-dropdown.jsx)
        -   [`ButtonCommandsList`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-commands-list.jsx)
    -   This use only focus:
        -   [`ButtonLinkAutocomplete`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link-autocomplete-list.jsx)
    -   These use nothing directly:
        -   [`ButtonTargetList`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-target-list.jsx)
        -   [`ButtonStylesList`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-styles-list.jsx)
        -   [`ButtonLinkEdit`](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-link-edit.jsx) (uses the `WidgetDropdown` HOC, which needs `moveFocus()`)

### `ToolbarButtons`

-   **Provides:**
    -   `cancelAnimation()`
    -   `getArrowBoxClasses()`
    -   `getConstrainedPosition()`
    -   `getInteractionPoint()`
    -   `getToolbarButtonGroups()`
    -   `getToolbarButtons()`
    -   `getWidgetXYPoint()`
    -   `isVisible()`
    -   `moveToPoint()`
    -   `show()`
    -   `updatePosition()`
-   **Requires:**
    -   `filterExclusive` (ie. from `WidgetExclusive`)
    -   `mergeExclusiveProps` (ie. from `WidgetExclusive`)
    -   `mergeDropdownProps` (ie. from `WidgetDropdown`)
-   **Examples:**
    -   [`ToolbarAdd`](https://github.com/liferay/alloy-editor/blob/master/src/components/toolbars/toolbar-add.jsx)
    -   [`ToolbarStyles`](https://github.com/liferay/alloy-editor/blob/master/src/components/toolbars/toolbar-styles.jsx)

# Using buttons from AlloyEditor 1.x in 2.x

To use custom 1.x buttons (created with `React.createClass`) in AlloyEditor 2.x, you need to make sure [a polyfill](https://www.npmjs.com/package/create-react-class) for `React.createClass` is available in your environment (I believe we already do this with liferay-portal), and you can access the old mixins as properties on the `AlloyEditor.Compat` object:

-   `AlloyEditor.Compat.ButtonActionStyle`
-   `AlloyEditor.Compat.ButtonCommandActive`
-   `AlloyEditor.Compat.ButtonCommand`
-   `AlloyEditor.Compat.ButtonKeystroke`
-   `AlloyEditor.Compat.ButtonProps`
-   `AlloyEditor.Compat.ButtonStateClasses`
-   `AlloyEditor.Compat.ButtonStyle`
-   `AlloyEditor.Compat.ToolbarButtons`
-   `AlloyEditor.Compat.WidgetArrowBox`
-   `AlloyEditor.Compat.WidgetDropdown`
-   `AlloyEditor.Compat.WidgetExclusive`
-   `AlloyEditor.Compat.WidgetFocusManager`
-   `AlloyEditor.Compat.WidgetInteractionPoint`
-   `AlloyEditor.Compat.WidgetPosition`

Note that the list corresponds to the one from 1.x, but `ButtonCfgProps` has been renamed to `ButtonProps` for consistency with the `ButtonProps` HOC from 2.x.

**Comment by mwilliams2014** [comment]
Thanks @wincent ! I'll review this and get back to you.

**Comment by antonio-ortega** [comment]
Hi @mwilliams2014 ,

Just as an additional note, the polyfill mention by @wincent in the last section was included in Liferay Portal in LPS-98752, so it is available for customers since:
 - fixpack-dxp-14 in 7.1
 - fixpack-de-86 in 7.0

Thanks!

**Comment by wincent** [comment]
@esanzp asked me to mirror this over to Google Drive, so it's over there too in "[Writing custom buttons with AlloyEditor](https://docs.google.com/document/d/1Mo0lRlBkVIPw9uroZBb_JUCNx2nJrsShzzmSbkDt7wY/edit)".
